src/gluonnlp/utils/parameter.py

```diff
@@ -180,3 +181,62 @@ def move_to_ctx(arr, ctx):
         return [move_to_ctx(ele, ctx) for ele in arr]
     else:
         return None if arr is None else arr.as_in_ctx(ctx)
+
+
+def deduplicate_param_dict(param_dict):
```

Comment:
@leezu @xinyual @szha  I added this `deduplicate_param_dict` because I find that `model.collect_params()` will return all parameters, even if some are shared. Thus, if there are shared parameters, `allreduce` will be called for multiple times. The results will thus be incorrect.

Reply:
@leezu shall we consider allowing deduplication in block.collect_params? we currently have regex as a filter already and enhancing filter logic with deduplication seems to make sense.